keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1228 @@
+
+[https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml) offers an API which appears to do what we need in both directions.
+
+### Header
+
+At the top of each KYAML document (or multi-document) will be the following header:
+
+```
+#!yaml
+---
+```
+
+This disambiguates it from JSON, since both KYAML and JSON begin with an opening curly-brace.  Kubernetes versions 1.33 and higher can handle KYAML without the header, but it is needed for versions before that.
+
+### Per-type rendering
+
+Integers and floats are rendered as their natural numeric representation.
+
+Simple strings are rendered as double-quoted and escaped. Multi-line strings require special consideration, see below.
+
+Booleans are rendered as `true` or `false`.
+
+Structs are rendered with curly-braces. Field names are not quoted unless they are not obviously strings (which includes prefixed label keys) or are one of the known type-ambiguous words (e.g. `no`).  Field values are rendered as per their type. Embedded structs will be rendered as inline fields.  Fields which have a `json` struct tag will be rendered according to that tag:
+  * the tag’s field name will be used
+  * if the tag’s name is “-”, the field will be skipped
+  * if the tag specifies “omitempty” it will be respected
+  * if the tag specifies “omitzero” it will be respected
+
+Maps are rendered with curly-braces. Keys are sorted and are not quoted unless they are type-ambiguous (e.g. `no`), and values are rendered as per their type.
+
+<<[UNRESOLVED quoted strings]>>
+Should we treat strings for field names and map keys:
+As described above, with label keys quoted but field names not
+Allow unquoted label keys and more complex (but not ambiguous) values
+Always quote strings, including ALL map keys and field names
+<<[/UNRESOLVED>>
+
+Slices and arrays are rendered with square-braces. Elements are rendered as per this list.
+
+Pointers are rendered as `null` or dereferenced and rendered as per the pointee’s type.
+
+Types which implement the Go standard `json.Marshaler` interface (`MarshalJSON()`) are rendered by marshaling to JSON, unmarshaling into an `any`, and then rendering that `any` as KYAML.
+
+Types which implement the `yaml.Marshaler` interface (`MarshalYAML()`) are rendered as per their type.  `MarshalYAML()` is not a Go standard and is not used by KYAML.
+
+<<[UNRESOLVED yaml self-marshaling]>>
+Should we do something for YAML self-marshalling?  It is non-standard and we do not use it in Kubernetes.
+<<[/UNRESOLVED>>
+
+Interfaces are rendered as `null` or their true value type.
+
+Any other type (as per `reflect.Value.Kind()`) will produce an error.
+
+### Multi-line strings
+
+For better or worse, multi-line strings are a thing that people do in Kubernetes, and conventional YAML makes it pretty easy. KYAML has to choose how we want to handle it.  Consider these examples of conventional YAML.
+
+Example 1:
```

Comment:
I think we neglected to mention that the spec doesn't permit this string format when in block style (braces etc), only flow style (indentation sensitive yaml).

Therefore these are not an option, they're incompatible with formatting in block style.

Reply:
Will note